### PR TITLE
Bootstrap 5 upgrade

### LINKED ISSUE #1190: Update OPUS to Bootstrap 5
Bootstrap 4 is reaching end of life in 11/2022. We should upgrade to Bootstrap 5.

### PR DESCRIPTION
**Pull request by dstopp**
- Fixes #1190
- Were any Django, import pipeline, or table_schema files modified? Y/N
  - If YES:
    - Database used: opus3_test_220530
    - All Django tests pass: NA
    - FLAKE8 run on modified code: NA
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: Y/NA
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): NA
- Was the documentation reviewed for necessary changes or additions? NA
  - About
  - Getting Started
  - FAQ
  - API Guide
  - Tooltips

Description of changes:

Known problems:
The close button on the modal still needs some adjusting; testing of changes is ongoing.  Testing still ongoing.

### CODE REVIEW

File: opus/application/.coveragerc [hunk 1]
```diff
@@ -1,12 +0,0 @@
-[run]
```
Comment (rfrenchseti) [inline opus/application/.coveragerc]: This file should not be removed

**Review by rfrenchseti — CHANGES_REQUESTED** [review]

**Comment by rfrenchseti** [comment]
A few observations:
- Sort pills don't have the right color text
- (x) to close dialogs doesn't work (as you know)
- Search menus don't have lines dividing them but have lines flash under the mouse as you move it around
- Select Metadata has no margin between "Download CSV (all results)" and the line below it

File: opus/application/apps/ui/templates/ui/base.html [hunk 2]
```diff
@@ -6,87 +6,89 @@
...
+                                    </div>
+                                </div>
+                                <a class="dropdown-item op-help-item" data-action="announcements" href="#" title="View recent announcements about changes to OPUS">Recent Announcements</a>
+                                <a class="dropdown-item op-help-item" data-action="volumes" href="#" title="View the list of PDS data volumes available in OPUS">Data Volumes</a>
+                                <a class="dropdown-item op-help-item" data-action="apiguide" href="#" title="View a description of the OPUS API for external developers">API Guide</a>
+                                <a class="dropdown-item op-help-item" data-action="splash" href="#" title="View the OPUS new-user splash screen">Welcome Message</a>
+                                <a class="dropdown-item op-help-item" data-action="citing" href="#" title="How to cite OPUS in your papers or presentations">Citing OPUS</a>
+                                <a class="dropdown-item op-help-item" data-action="contact" href="#" title="Send questions or feedback to the OPUS developers">Contact Us</a>
+                           </div>
```
Comment (juzen2003) [inline opus/application/apps/ui/templates/ui/base.html]: Add one more space to align with the opening <div>

**Review by juzen2003 — CHANGES_REQUESTED** [review]
- The pre-programmed input dropdown is broken (ex: wavelength widget). The preprogrammed dropdown will not show up when clicking on the input.
- When opening up the download links history (at the lower right corner), the dialog window has missing (x) button and "Clear All" button. 
- When clicking on the "trash" icon in the header of any column in the browse table view, the column will be removed but the following error message will show up in the console.
```
Uncaught TypeError: Failed to execute 'getComputedStyle' on 'Window': parameter 1 is not of type 'Element'.
```

File: opus/application/static_media/css/opus.css [hunk 3]
```diff
@@ -1233,14 +1289,23 @@ without resizing image proportionally. */
     padding-right: 0.5em;
 }
 
-#op-metadata-detail-view-content div.close {
+/* Make sure the right part of the slide modal will take proper space when slide
+modal is narrow */
+.op-remove-col {
+    flex: 1 0 auto !important;
+    width: 100%
```
Comment (juzen2003) [inline opus/application/static_media/css/opus.css]: missing ```;``` at the end.

File: opus/application/static_media/css/opus.css [hunk 4]
```diff
@@ -1910,17 +1988,27 @@ panel are properly separated */
...
     height: .1em;
+    font-size: 1.1em;
+    font-family: Verdana,Arial,sans-serif;
+}
+
+.op-observation-slider .ui-state-focus {
+    outline: none;
+    background: inherit;
+    border: 1px solid #c5c5c5;
+    color:
```
Comment (juzen2003) [inline opus/application/static_media/css/opus.css]: ```color``` styling is not used.

File: opus/application/static_media/js/browse.js [hunk 5]
```diff
@@ -1876,7 +1885,7 @@ var o_browse = {
         opus.colLabelsNoUnits = columnsNoUnits;
 
         // check all box
-        // let addallIcon = "<button type='button' data-toggle='modal' data-target='#op-addall-to-cart-modal' " +
+        // let addallIcon = "<button type='button' data-bs-toggle='modal' data-bs-target='#op-addall-to-cart-modal' " +
```
Comment (juzen2003) [inline opus/application/static_media/js/browse.js]: Unused code.

**Comment by rfrenchseti** [comment]
- The Help button, when selected, has an action when you mouse over it (the text turns white). This doesn't happen anywhere else. Is this intentional?
- "Download Links History" in the bottom right is up against the right edge of the browse. Currently there is whitespace.

**Comment by juzen2003** [comment]
- The ```Download CSV (all results)``` button in select metadata menu will stay at the same location when the select metadata menu window is moved. This happened at current master (production site) as well.
-  Preprogrammed ranges works now but the console error message is still there, I'll dig in into it now.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
A few of these are cosmetic and may not be worth fixing (or we like the new versions better) but many of the cosmetic changes I think look worse than the old version. Others are just plain bugs.

- On search widgets, the vertical bar to the left of the minimize/close buttons is now very pale.
- On the other hand, on search widgets that have an "-OR-", the "-" beside the OR are now dark black instead of the same grey as the word "OR".
- On search widgets, the title bar is not a pure grey instead of blue-grey
- On a preprogrammed ranges dropdown, the headers are now in grey text, which makes them harder to read. The text under each header is also grey.
- The headings under the Help menu are grey
- When clicking on Interpreting Preview Images under Help, the old behavior was animated and the new behavior just pops into existence
- When the browser is shrunk, the Reset to Default button in Select Metadata becomes misplaced
- When you add a sort field by pressing "+" next to "Sort by", the main help menu pops up in a weird place and never goes away
- As you shrink the browser vertically, you can make the slideshow dialog move off the top of the screen
- The hamburger menu entries on the Browse tab are grey
- When in table view, if you start a range with Shift+click or "Start range" from a hamburger menu, the green animated square is in the wrong place and the checkbox and hamburger move down some
- When in table view with enough columns that you need a horizontal scrollbar, if you scroll to the right and then show the "+ Trash +" dropdown, it's in the wrong place
- There is a very slight gap above the table header in table view such that if you scroll up the data you can see parts of white characters in the gap
- If you activate the hamburger menu, then click the menu's titlebar to close it, the main Help menu shows up and won't go away
- If you pull up the Detail tab for one of the observations with multiple preview images (e.g. UVIS/VIMS/RSS occultation profiles), no previews are shown at all. This might be a merge problem.
- I see the following error in the console whenever I switch to the Detail tab:

```
jquery-3.6.0.min.js:2 jQuery.Deferred exception: $(...).flexslider is not a function TypeError: $(...).flexslider is not a function
    at Array.<anonymous> (http://127.0.0.1:8000/static_media/js/detail.js?version=5c7f916830a2c3fac2966bca03e1ab8f4d84a0f5:135:42)
    at e (https://code.jquery.com/jquery-3.6.0.min.js:2:30038)
    at t (https://code.jquery.com/jquery-3.6.0.min.js:2:30340) undefined
```

- There is not as much vertical margin between the horizontal blue divider at the top of the screen and the View Table line (browse or cart) as there used to be. I think it looks asymmetric and cramped now.

File: opus/application/static_media/css/opus.css [hunk 6]
```diff
@@ -40,8 +43,54 @@ body.modal-open .navbar {
...
+.no-gutters > .col,
+.no-gutters > [class*="col-"] {
+  padding-right: 0;
+  padding-left: 0;
+}
+
+
 /* Make the main nav bar just a little taller than normal (style says 8px) */
-#op-main-nav {
+#op-main-nav  {
```
Comment (rfrenchseti) [inline opus/application/static_media/css/opus.css]: Extra space?

File: opus/application/static_media/js/browse.js [hunk 7]
```diff
@@ -2729,15 +2737,15 @@ var o_browse = {
         let tab = opus.getViewTab();
         let targetElem = e.currentTarget;
         let toolElem = `${tab} .op-edit-field-tool`;
-        let left = $(targetElem).position().left +
-                    $(`${tab} .op-data-table-view .table`).offset().left;
+        let left = $(targetElem).position().left ;/*+
```
Comment (rfrenchseti) [inline opus/application/static_media/js/browse.js]: Why is this commented out?

File: opus/application/static_media/js/search.js [hunk 8]
```diff
@@ -276,13 +266,15 @@ var o_search = {
...
                 // close the dropdown
                 let inputToTriggerDropdown = $(`#widget__${slugName} input.op-range-input-min[name="${inputName}"]`);
-                let preprogrammedRangesDropdown = (inputToTriggerDropdown
-                                                   .next(".op-preprogrammed-ranges")
-                                                   .find(".op-scrollable-menu"));
+                let preprogrammedRangesDropdown = inputToTriggerDropdown.next(".op-preprogrammed-ranges");
 
                 if (preprogrammedRangesDropdown.hasClass("show")) {
-                    inputToTriggerDropdown.dropdown("toggle");
+                    // preprogrammedRangesDropdown.toggleClass("show");
```
Comment (rfrenchseti) [inline opus/application/static_media/js/search.js]: Remove comments?

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
Shouldn't you remove infinite_scroll version 3.0.5? Same for old versions of jquery.

File: opus/application/static_media/css/opus.css (hunk 6 above)
Comment (dstopp) [reply opus/application/static_media/css/opus.css]: I removed the extra space; re: infinite_scroll version 3.0.5, that is the current version we are testing with.  4.0.1 is not yet implemented and should be done after the bs5 work is complete.
re: jquery, fontawesome, popper and lodash - removed unused/outdated files.

File: opus/application/static_media/js/browse.js (hunk 7 above)
Comment (dstopp) [reply opus/application/static_media/js/browse.js]: removed commented out code

File: opus/application/static_media/js/search.js (hunk 8 above)
Comment (dstopp) [reply opus/application/static_media/js/search.js]: removed commented out code

File: opus/application/static_media/js/browse.js [hunk 9]
```diff
@@ -2729,15 +2737,14 @@ var o_browse = {
...
+        let left = $(targetElem).position().left;
         let top = $(targetElem).offset().top +
                   $(targetElem).outerHeight();
 
         if (tab === "#cart") {
-            if ($(".cart_details:visible").length > 0) {
+            if ($(".op-cart-details:visible").length > 0) {
                 // need to calculate a new offset if the table has scrolled left behind the details window
-                left -= $(".cart_details").outerWidth() + $(".cart_details").offset().left;
+                //left -= $(".op-cart-details").outerWidth() + $(".op-cart-details").offset().left;
```
Comment (juzen2003) [inline opus/application/static_media/js/browse.js]: This if statement block can be removed since the code is commented out.

**Review by juzen2003 — CHANGES_REQUESTED** [review]
When clicking the items under the following two dropdown menus, there will be error messages in the console from dropdown.js (from bootstrap)
- "+" to add a new field in "sort by" row
- "+ Add new field here" to add a new field in browse view metadata modal 
The functionality is good, just the error messages will always show up when an item is selected in either of the menus.

**Comment by rfrenchseti** [comment]
If you activate the hamburger menu on an observation (like a thumbnail) and then select one of the menu items (like add to cart), an error shows up in the console.

```
dropdown.js:241 Uncaught TypeError: Cannot read properties of null (reading 'classList')
    at li._isShown (dropdown.js:241:23)
    at li.toggle (dropdown.js:117:17)
    at HTMLAnchorElement.<anonymous> (dropdown.js:436:38)
    at HTMLDocument.n (event-handler.js:118:19)
```

**Comment by rfrenchseti** [comment]
If you create a download file from the cart page, the popup box that gives the list of URLs for downloads shows up where the mouse is, not in the lower right. If you close it and try to reopen it, it won't stay open and the spinning-busy-symbol is shown in the box for the brief time you can see it.

File: opus/application/static_media/js/browse.js (hunk 9 above)
Comment (rfrenchseti) [reply opus/application/static_media/js/browse.js]: It doesn't look like this comment was ever acted on

File: opus/application/test_api/test_results_api.py [hunk 10]
```diff
@@ -385,7 +385,7 @@ def test__api_metadata_vg_iss_2_s_c4360845_default_html(self):
     def test__api_metadata_vg_iss_2_s_c4360845_default_html_private(self):
         "[test_results_api.py] /api/metadata_v2: vg-iss-2-s-c4360845 default html private"
         url = '/__api/metadata/vg-iss-2-s-c4360845.html'
-        expected = b'<ul id="detail__data_general_constraints" class="op-detail-list">\n<li class="op-detail-entry">\n<i class="fas fa-info-circle op-detail-entry-icon op-detail-metadata-tooltip" data-toggle="tooltip"\ntitle="'
+        expected = b'<ul id="detail__data_general_constraints" class="op-detail-list">\n<li class="op-detail-entry">\n<i class="fas fa-info-circle op-detail-entry-icon op-detail-metadata-tooltip" data-bs-toggle="tooltip"\ntitle="'
```
Comment (rfrenchseti) [inline opus/application/test_api/test_results_api.py]: Bad merge

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
- In the Help menu, the submenus for Interpreting Preview Images don't link to anything when you click on them.
- Bootstrap 5.2.2 is the current version but we are using 5.2.0.
- opus.css has inconsistent whitespace - some indents are 2 chars and some are 4.
- search.js and widgets.js use the symbol `bootstrap` but it wasn't declared as a global at the top of the file, thus causing errors in JSHINT
- ui/templates/ui/header.html still has references to bootstrap 5.1.3. This part of the file is used when creating PDF versions of help files, and this is why PDF creation is currently failing.
- Pull up an observation in the slideshow. Use the edit mode to add a metadata field. The UI will freeze completely. Or sometimes it will refresh the display but not actually let you add a field. This only happens from the browse tab, not the cart tab.
- When you "start add range to cart" from the table view, there is no green box indicator on the left

**Comment by rfrenchseti** [comment]
When in the table view, if you delete a column by clicking on the trashcan icon you get the following in the console:

Uncaught TypeError: this._menu is null
    _isShown dropdown.js:241
    toggle dropdown.js:117
    <anonymous> dropdown.js:436
    n event-handler.js:118

**Comment by rfrenchseti** [comment]
This is actually a problem in the production OPUS too:
If the "Download Links History" box is open, and the browser is shrunk very small, the footer goes away. But when it goes away, that links history box becomes detached from the footer area and shows up in the top left, partially off the screen.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
- The current master branch is not merged in.
- 6 tests are failing.
- Can you just remove the infinite-scroll-4.0.1.pkg since we aren't using it yet?
- I was still able to get the slideshow add/remove metadata to freeze up, but it's harder to do now. We can defer this to a future pull.
- HOWEVER, with the current version, delete a column while in table view always freezes now. This didn't happen before.

**Review by juzen2003 — CHANGES_REQUESTED** [review]
- Add multiple observations to cart, click view gallery & view table, this console error message will show up whenever we switch between these two views in cart:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'backdrop')
```
- In search tab, mult values are duplicated for the Ground-based category in instrument name widget, this can be fixed by merging current master branch.

**Review by rfrenchseti — APPROVED** [review]

**Review by juzen2003 — APPROVED** [review]
- Database used: re-import
- All Django tests pass: Y
- JSHINT run on all affected files: NA
- FLAKE8 run on all affected files: Y
- Code coverage run: Y
- Tested on Chrome, Firefox, Safari, and iOS: NA
  (Remember to test all browser sizes)
- Tested for race conditions (with delayed API calls if applicable): NA

**Pull request merged**